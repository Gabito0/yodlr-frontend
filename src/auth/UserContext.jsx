import React from "react";

/**
 * UserContext provides a context for user information in the application.
 *
 * This context allows different parts of the app to access and update the current
 * user's information, such as their ID, email, and role (e.g., admin status).
 * It facilitates state management for user data across various components.
 *
 * Components can consume this context to access user data or to make updates
 * when user information changes, such as during login, logout, or profile updates.
 */

const UserContext = React.createContext();

export default UserContext;
