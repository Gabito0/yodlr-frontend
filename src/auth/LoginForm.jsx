import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import AppAlert from "../common/AppAlert";
import UserContext from "./UserContext";

/**
 * LoginForm component renders a login form.
 *
 * It manages the state of the form data and handles the form submission.
 * It redirects the user to the profile page upon successful login.
 */
const LoginForm = ({ login }) => {
  const navigate = useNavigate(); // Hook for navigation
  const { currentUser } = useContext(UserContext); // Get current user from context

  // State for form data and form errors
  const [formData, setFormData] = useState({
    email: "",
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  // Redirect to profile if already logged in
  useEffect(() => {
    if (currentUser) {
      navigate(`/profile`);
    }
  }, [currentUser, navigate]);

  // console.debug(
  //   "LoginForm",
  //   "login=",
  //   typeof login,
  //   "formData=",
  //   formData,
  //   "formErrors=",
  //   formErrors
  // );

  // Handle form submission
  async function handleSubmit(evt) {
    evt.preventDefault();
    const result = await login(formData);
    if (result.success) {
      navigate("/profile"); // Navigate to profile on successful login
    } else {
      setFormErrors(result.errors); // Set form errors on failure
    }
  }

  // Update form data state on input change
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((data) => ({ ...data, [name]: value }));
  }

  return (
    <div className="LoginForm">
      <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
        <h3 className="mb-3">Log In</h3>
        <div className="card">
          <div className="card-body">
            <form onSubmit={handleSubmit}>
              <div className="form-group">
                <label htmlFor="email">Email</label>
                <input
                  type="email"
                  name="email"
                  id="email"
                  className="form-control"
                  value={formData.email}
                  onChange={handleChange}
                  autoComplete="email"
                  required
                />
              </div>
              <div className="form-group">
                <label htmlFor="password">Password</label>
                <input
                  type="password"
                  name="password"
                  id="password"
                  className="form-control"
                  value={formData.password}
                  onChange={handleChange}
                  autoComplete="current-password"
                  required
                />
              </div>
              {formErrors.length > 0 && (
                <AppAlert type="danger" messages={formErrors} />
              )}
              <button className="btn btn-primary float-right mt-3">
                Submit
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default LoginForm;
