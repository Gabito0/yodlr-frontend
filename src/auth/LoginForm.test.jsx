import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import LoginForm from "./LoginForm";
import UserContext from "./UserContext";
import { vi, expect, it } from "vitest";

// Mock the login function
vi.mock("../api/api");

it("matches snapshot", function () {
  const mockLogin = vi.fn();
  const mockCurrentUser = null;

  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ currentUser: mockCurrentUser }}>
        <LoginForm login={mockLogin} />
      </UserContext.Provider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

it("submits form with valid data", async function () {
  const mockLogin = vi.fn().mockResolvedValue({ success: true });

  const { getByLabelText, getByText } = render(
    <MemoryRouter>
      <UserContext.Provider value={{ currentUser: null }}>
        <LoginForm login={mockLogin} />
      </UserContext.Provider>
    </MemoryRouter>
  );

  fireEvent.change(getByLabelText("Email"), {
    target: { value: "test@example.com" },
  });
  fireEvent.change(getByLabelText("Password"), {
    target: { value: "password" },
  });

  fireEvent.click(getByText("Submit"));

  await waitFor(() => {
    expect(mockLogin).toHaveBeenCalledWith({
      email: "test@example.com",
      password: "password",
    });
    expect(mockLogin).toHaveBeenCalledTimes(1);
  });
});
