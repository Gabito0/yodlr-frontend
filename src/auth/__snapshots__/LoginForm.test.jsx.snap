// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="LoginForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h3
        class="mb-3"
      >
        Log In
      </h3>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="form-control"
                id="email"
                name="email"
                required=""
                type="email"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="form-control"
                id="password"
                name="password"
                required=""
                type="password"
                value=""
              />
            </div>
            <button
              class="btn btn-primary float-right mt-3"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
